oauthenticator/globus.py

```diff
@@ -72,22 +43,58 @@ class GlobusLoginHandler(OAuthLoginHandler, GlobusMixin):
     def get(self):
         redirect_uri = self.authenticator.get_callback_url(self)
         client = self.authenticator.globus_portal_client()
+        state = self.get_state()
+        self.set_state_cookie(state)
         client.oauth2_start_flow(
             redirect_uri,
             requested_scopes=' '.join(self.authenticator.scope),
-            refresh_tokens=self.authenticator.allow_refresh_tokens
+            refresh_tokens=self.authenticator.allow_refresh_tokens,
+            state=state
         )
         self.log.info('globus redirect: %r', redirect_uri)
         self.authorize_redirect(client)
 
 
+class GlobusLogoutHandler(LogoutHandler):
+    """
+    Handle custom logout URLs and token revocation. If a custom logout url
+    is specified, the 'logout' button will log the user out of that identity
+    provider in addition to clearing the session with Jupyterhub, otherwise
+    only the Jupyterhub session is cleared.
+    """
+    @gen.coroutine
+    def get(self):
+        user = self.get_current_user()
+        if user:
+            if self.authenticator.revoke_tokens_on_logout:
+                self.clear_tokens(user)
+            self.clear_login_cookie()
+        if self.authenticator.logout_redirect_url:
+            self.redirect(self.authenticator.logout_redirect_url)
+        else:
+            super().get()
```

Comment:
The problem is the `self.redirect()` in the base [Jupyterhub Logout Handler](https://github.com/jupyterhub/jupyterhub/blob/master/jupyterhub/handlers/login.py#L25). We can't do a redirect after a redirect, so this code explicitly checks to make sure that doesn't happen. An alternative could be adding an issue to Jupyterhub to make the logout URL configurable:

    class LogoutHandler(BaseHandler):
        self.redirect_url = self.settings['login_url']
        def get(self):
            ...
            self.redirect(self.redirect_url, permanent=False)

Also, since the addition of `auth_state` to the authenticator, I needed to add some duplicate code to grab the user again to get at `auth_state` to revoke tokens. It's possible other services such as MediaWiki may want to do something to `auth_state` on logout, where the following addition would be convenient: 

    class LogoutHandler(BaseHandler):
        def get(self):
            user = self.get_current_user()
            if user:
                post_user_logout(user)
            ...
        def post_user_logout(self, user):
            """Override this to revoke user tokens in auth_state"""
            pass

I can add those as separate Jupyterhub issues if you think they merit the discussion. Unfortunately for the current code above, I don't think `super().get()` can be moved due to the second redirect.

Reply:
> I can add those as separate Jupyterhub issues if you think they merit the discussion. 

Please do, I think both make sense as upstream issues to be handled.

Reply:
Great! I'll add them.